sql/database.go

```diff
@@ -772,6 +819,30 @@ func (db *sqliteDatabase) Close() error {
 	return nil
 }
 
+// Connection implements Database.
+func (db *sqliteDatabase) Connection(ctx context.Context) (Connection, error) {
+	if db.closed {
+		return nil, ErrClosed
+	}
+	conCtx, cancel := context.WithCancel(ctx)
+	conn := db.getConn(conCtx)
+	if conn == nil {
+		cancel()
+		return nil, ErrNoConnection
+	}
+	return &sqliteConn{queryCache: db.queryCache, db: db, conn: conn, freeConn: cancel}, nil
+}
```

Comment:
I don't think that a `context.Context` needs to be passed here. With it there are 2 identical ways to release a connection: calling `cancel()` on the passed context or calling `Release` on the returned connection. Instead I think it would be sufficient to only have the `Release` way to return a connection to the pool:

```suggestion
func (db *sqliteDatabase) Connection() (Connection, error) {
	if db.closed {
		return nil, ErrClosed
	}
	conCtx, cancel := context.WithCancel(context.Background)
	conn := db.getConn(conCtx)
	if conn == nil {
		cancel()
		return nil, ErrNoConnection
	}
	return &sqliteConn{queryCache: db.queryCache, db: db, conn: conn, freeConn: cancel}, nil
}
```

Reply:
There's a following problem with this approach.
Let's assume we have some func that gets a `Context` and passes it to the `Connection`:
```go
func foo(ctx context.Context, db sql.Database) {
  conn := db.Connection(ctx)
  defer conn.Release()
  // do a potentially long-running query
  if _, err := sql.Exec(...) {
    return fmt.Errorf("...", err)
  }
  ...
  return nil
}
```
This way, when `ctx` is canceled, the query is interrupted immediately due to `SetInterrupt` in crawshaw library.
If we remove the `ctx` argument from `Connection`, the query will always have to run to its completion when the context is canceled, or we would have to start another goroutine to monitor the context and then call `conn.Release()`, which is some extra error-prone burden.

Reply:
To me its a code smell that cancelling the context and calling `conn.Release()` are functionally identical. To me this means that one of the two shouldn't be needed: Either the connection is released via context cancellation:

```go
func foo(ctx context.Context, db sql.Database) error {
    ctx, cancel := context.WithCancel(ctx)
    defer cancel()
    conn := db.Connection(ctx)
    // one or multiple queries
}
```

or via a deferred `Release()` call:

```go
func foo(ctx context.Context, db sql.Database) error {
    conn := db.Connection()
    defer conn.Release()
    // one or multiple queries
}
```

I prefer the second approach because the first is not very intuitive and can lead to objects living longer than intended (e.g. if `ctx` is passed without wrapping it with `context.WithCancel` first). Generally a context should be discardable after the function that it was passed to returns, i.e. I can call `cancel` immediately without any effects.

I know the second approach has the disadvantage that `sql.Exec` won't return immediately when `ctx` is cancelled, but why should it? I would only expect it to do so when it gets the `ctx` passed directly as argument.

The current implementation is not intuitive: cancelling the `ctx` that was passed to `db.Connection` causes `sql.Exec` to fail. Instead if we want or need the quick exit we should try to pass `ctx` to exec. Then its clear from the code that `sql.Exec` can fail with `context.Canceled` (wrapped with more info). And in this case we still don't need to pass a `ctx` to `db.Connection`. 🙂

Reply:
`Connection` is modeled after `Transaction`. When we initiate a transaction, we have to pass the context, and a long-running SQL query may be interrupted, as the crawshaw library uses the context to establish interrupt handler on the connection when it's retrieved from the pool. There's no possibility in crawshaw to interrupt a single command without invoking the interrupt handler on the connection.
`Connection` also retrieves the connection from the pool and so it must pass the context here.
Another problem with passing ctx to `Exec` is that we'll need to change our `sql` package soon to only accept context for read transactions (and connections). When a write transaction is started, interrupting it by canceling context may lead to deadlocks, SQLite+crawshaw just don't handle that properly. Thus, we'll have `ctx` argument in `Transaction` / `WithTransaction` / `Connection` / `WithConnection`, but not in `TxImmediate` or `WithTxImmediate`; and all the conns in the pool that are used by `Transaction` / `WithTransaction` / `Connection` / `WithConnection` will be read-only, with just a single connection being available for TxImmediate` or `WithTxImmediate`. See #6273 

Re just not interrupting the queries: some queries may take time to complete, and it would make sense to stop them right away upon context cancelation if they don't do any writes; and syncv2 doesn't write to the database by itself, relying on the handlers for that.

Reply:
It makes no sense to pass a context to a function that returns (almost) immediately but still holds on to that context. That is just bad code design. Any context should be cancellable immediately after a function that received it returns otherwise unexpected side effects can happen:

```go
func (s *Service) GetConnection(ctx context.Context) sql.Connection {
    conCtx := context.WithTimeout(ctx, 5*time.Second)
    defer cancel()  // this must be deferred to stop a go routine that was created with the context
    conn := db.Connection(conCtx)
    // do something else
    return conn // connection will be closed latest after 5 seconds - which is NOT what I would expect here - the context was for "creating" the connection
}
```

Reply:
Furthermore I don't think this is an issue of crawshaw, but rather how we use that library 🙂

Reply:
Removed `Database.Connection`, kept just `Database.WithConnection`.
Updated #6446, #6447 and #6448 accordingly.